libraries/type-length-value/src/state.rs

```diff
@@ -408,6 +408,54 @@ impl<'a> TlvState for TlvStateMut<'a> {
     }
 }
 
+/// Packs a borsh-serializable value into an existing TLV space, reallocating
+/// the account and TLV as needed to accommodate for any change in space
+#[cfg(feature = "borsh")]
+pub fn realloc_and_borsh_serialize<V: TlvDiscriminator + borsh::BorshSerialize>(
+    account_info: &solana_program::account_info::AccountInfo,
+    value: &V,
+) -> Result<(), ProgramError> {
+    let previous_length = {
+        let data = account_info.try_borrow_data()?;
+        let TlvIndices {
+            type_start: _,
+            length_start,
+            value_start,
+        } = get_indices(&data, V::TLV_DISCRIMINATOR, false)?;
+        usize::try_from(*pod_from_bytes::<Length>(&data[length_start..value_start])?)?
+    };
+    let new_length = solana_program::borsh::get_instance_packed_len(&value)?;
+    let previous_account_size = account_info.try_data_len()?;
+    if previous_length < new_length {
+        // size increased, so realloc the account, then the TLV entry, then write data
+        let additional_bytes = new_length
+            .checked_sub(previous_length)
+            .ok_or(ProgramError::AccountDataTooSmall)?;
+        account_info.realloc(previous_account_size.saturating_add(additional_bytes), true)?;
+        let mut buffer = account_info.try_borrow_mut_data()?;
+        let mut state = TlvStateMut::unpack(&mut buffer)?;
+        state.realloc::<V>(new_length)?;
+        state.borsh_serialize(value)?;
```

Comment:
You're going to need to transfer lamports in or out based on the new size (increased size will error if no new lamports are added for rent)

Seems like you're putting the onus on the developer to take this step, which also bleeds into my point in my other comment about this maybe not being necessary to do with `AccountInfo` directly

Reply:
That's correct, I don't want to include payers anywhere in interfaces, since that provides easy attack opportunities for programs that implement the interface.

For best use, a program should implement some `withdraw_excess_lamports` interface for the right authority to claim additional lamports, and the tools need to transfer the additional required lamports as a separate instruction, as demonstrated in the tests.

If you have a better idea of where this could go, I'm all ears! I definitely took the easy way out here. We can add a `borsh.rs` or `program.rs` file with borsh / solana program-specific utilities?